feat(containerize): Add support for macOS

dcarley opened the pull request:
## Proposed Changes

By using the proxy container approach from:

- https://www.notion.so/floxdev/012-Containerize-cross-platform-14a744c8e3168090b508dafd1fc9f37d0

This has some rough edges which we'll polish later..

The proxy container has to fetch/build store paths on every run. We can
reduce this time by caching/persisting the proxy container in #2469. It
may also be desirable to separate the build of the Flox flake from the
build of the environment for better feedback, but it's not worth doing
that until we refactor for caching.

Testing has only been done manually so far because we don't have a
functioning Docker or Podman in CI. This will be covered in #2466.

Output from `nix run` interleaves the "Writing container" spinner but it
does give some useful feedback about why it might be slow and is
consistent with the existing behaviour when writing layers. I've created
#2498 to put this all behind a verbose flag.

I've tested that it also works for remote environments.

Creating a `ContainerSource` for macOS doesn't require a spinner because
it doesn't execute anything until `stream_container()`.

Example output: https://gist.github.com/dcarley/99705356f0060920972dbf11ff5df812

## Release Notes

`flox containerize` now supports running from macOS!

ysndr commented on cli/flox/src/commands/containerize/macos_containerize_proxy.rs:
```diff
@@ -1,21 +1,104 @@
         &self,
-        _name: impl AsRef<str>,
-        _tag: impl AsRef<str>,
+        _name: impl AsRef<str>, // Inferred from `self.environment_path`
```
```suggestion
        _name: impl AsRef<str>, // Inferred from `self.environment_path`
                                // by flox _inside_ the container.
```

dcarley replied:
Done (will rebase fixup before merge)

ysndr commented on cli/flox/src/commands/containerize/macos_containerize_proxy.rs:
```diff
@@ -1,21 +1,104 @@
+        tag: impl AsRef<str>,
     ) -> Result<ContainerSource, Self::Error> {
-        todo!("🚧 MacOS container builder in construction 🚧")
+        let env_mount = "/flox_env";
+        let home_mount = "/flox_home";
```
suggestion: define as constants

dcarley replied:
Done (will rebase fixup before merge)

ysndr commented on cli/flox/src/commands/containerize/macos_containerize_proxy.rs:
```diff
@@ -1,21 +1,104 @@
+                ),
+            ]);
+        }
+
```
suggestion: disable metrics iff disabled by the host

dcarley replied:
Done (will rebase fixup before merge)

dcarley commented on cli/flox/src/commands/containerize/macos_containerize_proxy.rs:
```diff
@@ -1,21 +1,104 @@
+const FLOX_PROXY_IMAGE: &str = "ghcr.io/flox/flox";
+
+#[derive(Debug, Clone)]
+pub(crate) struct ContainerizeProxy {
```
I'll add some docstrings for this.

dcarley replied:
Done (will rebase fixup before merge)

dcarley commented on cli/flox/doc/flox-containerize.md:
```diff
@@ -46,10 +50,6 @@ Running the container non-interactively with `docker run <container id>`
 allows you to run a command within the container without launching a subshell,
 similar to `flox activate --`.
 
-**Note**:
-The `containerize` command is currently **only available on Linux**.
-The produced container however can also run on macOS.
```
I also need to remove the warning from https://flox.dev/docs/tutorials/sharing-environments/?h=container#sharing-with-containers and might as well do https://github.com/flox/product/issues/767 at the same time.

ysndr commented on cli/flox/src/commands/containerize/macos_containerize_proxy.rs:
```diff
@@ -70,6 +71,14 @@ impl ContainerBuilder for ContainerizeProxy {
             ]);
         }
 
+        // Honour FLOX_DISABLE_METRICS, if set.
+        if let Ok(disable_metrics) = std::env::var(FLOX_DISABLE_METRICS_VAR) {
+            command.args([
+                "--env",
+                &format!("{}={}", FLOX_DISABLE_METRICS_VAR, disable_metrics),
+            ]);
+        }
+
```
maybe the value should indeed come from the config, as it is if you disable metrics through the system-wide config file in `/etc/`, they will be _enabled_ in the container.
alternatively we might want to mount the etc config file explicitly.

ysndr replied:
looks like we set this env var earlier, maybe worth reminding us of that

dcarley replied:
👍🏻 

>         // Honour `FLOX_DISABLE_METRICS` if set. Aside from being set by the
>         // user, it may also be set at runtime by  [Flox::Commands::FloxArgs]
>         // from another config path like `/etc/flox.toml` which isn't mounted
>         // into the proxy container.

dcarley commented:
@ysndr could you take another look at this? Changes since last review are:
- rebased in previous feedback fixups
- removed the fallback to a specific commit now that v1.3.7 is released
- created #2502 because the commit precedence doesn't work from devShell or integration tests
- created #2498 to improve output handling
- haven't moved it to the provider because I realised that it depends on `Runtime` from CLI and we should ship as is

mkenigs commented on cli/flox/src/commands/containerize/macos_containerize_proxy.rs:
```diff
@@ -1,21 +1,115 @@
+use crate::config::FLOX_DISABLE_METRICS_VAR;
+
+const FLOX_FLAKE: &str = "github:flox/flox";
+const FLOX_PROXY_IMAGE: &str = "ghcr.io/flox/flox";
```
How come `ghcr.io/flox/flox` instead of `nixos/nix`?

dcarley replied:
It's described a bit further down:

>         // Use a released Flox container of the same semantic version as a base
>         // because it already has:
>         //
>         // - most of the dependency store paths
>         // - substitutors configured
>         // - correct version of nix
>         let flox_container = format!("{}:{}", FLOX_PROXY_IMAGE, flox_version_tag);
>         command.arg(flox_container);

It could be swapped for a `nixos/nix` container in or after https://github.com/flox/flox/issues/2469 if we're able to satisfy all of those.
